clang/lib/Driver/Driver.cpp

```diff
@@ -2708,7 +2708,11 @@ int Driver::ExecuteCompilation(
     // Remove result files if we're not saving temps.
     if (!isSaveTempsEnabled()) {
       const JobAction *JA = cast<JobAction>(&FailingCommand->getSource());
-      C.CleanupFileMap(C.getResultFiles(), JA, true);
+      // When performing offload compilations, the result files may not match
+      // the JobAction that fails.  In that case, do not pass in the JobAction
```

Comment:
> // When performing offload compilations, the result files may not match the JobAction that fails.

Can you add some details on why they don't match?
And also given an example Offload JobAction for this scenario?

Reply:
The generated result files are associated with a specific `JobAction`.  One example here is with the old offloading model.  Here, the output object from `clang++ -fsycl file.cpp -o file.o` is generated from the `clang-offload-bundler` call.  Both the host and device compilations have temporary output files, so the `JobAction` associated with the actual fail (if one fails) including the output file do not match.

Here, the failing action is a `BackendCompileJobAction` but the file that needs to be removed is from the `OffloadBundlingJobAction`

Reply:
> Here, the output object from clang++ -fsycl file.cpp -o file.o is generated from the clang-offload-bundler call.

`clang++ -fsycl file.cpp -o file.o`  will trigger both compilation and linking at the same invocation and will not call clang-offload-bundler.

AFAIK, when compilation is separated from linking, the compilation step ends with engaging the offload bundler to generate  "fat object" (<host object, device code IR> pair).

> Both the host and device compilations have temporary output files, so the JobAction associated with the actual fail (if one fails) including the output file do not match

Can you clarify this?
If host or device compilation fails, there will be no host object or device object file generated right?

Reply:
Oh - yeah `-c` is needed in that compilation line.  Thanks for pointing that out.

> Can you clarify this?
> If host or device compilation fails, there will be no host object or device object file generated right?

Yes, that is right, no object will be generated.  The issue is an existing object that needs to be cleaned up.  During the object compilation, the compiler is responsible for cleaning up the output file upon failure.  Since the output file in those instances are temporary - there is no association with the actual output file pointed to with `-o`.  The driver knows what the final output object file is, so we need to take the steps to clean up upon failure.

Reply:
Did I understand it correctly that if the host and device object files have been created, but the `clang-offload-bundler` fails to produce the fat object, we should still proceed with deleting the host and device object files?
In this case, the failing JobAction is OffloadBundlingJobAction but the host and device object files were created by a different JobAction.

Is is not possible to pass the JobAction associated with generating the host and device object files instead of not passing any JA at all?

Reply:
> the actual output file we want to remove

What is the actual output file we want to remove in the case where  `OffloadBundlingJobAction` fails?
My understanding is we want to remove the host and device object files, since those will be the only ones created.

Reply:
If the `OffloadBundlingJobAction` fails, we still want to remove the resulting output file, which is the 'fat' object.  This is the named object `file.o` from `clang++ -c -fsycl file.cpp -o file.o`

Reply:
> If the OffloadBundlingJobAction fails, we still want to remove the resulting output file

I'm confused. You mentioned above that if a compilation fails, there will be no result/output file generated for that failed compilation.
So if `OffloadBundlingJobAction` fails, there will be no fat object created right?

Reply:
> So if OffloadBundlingJobAction fails, there will be no fat object created right?

That's right.  But we still need to remove the object if one existed before the compilation.  This allows for matching behavior between `clang++ -c a.cpp -o a.o` and `clang++ -fsycl -c a.cpp -o a.o` if compilation fails in any way.